src/effects/effectchainslot.cpp

```diff
@@ -114,12 +144,408 @@ EffectChainSlot::~EffectChainSlot() {
+        m_effects.append(pEffect);
+        return;
+    }
+
+    VERIFY_OR_DEBUG_ASSERT(!m_effects.contains(pEffect)) {
+        return;
+    }
+
+    m_effects.append(pEffect);
+    if (m_bAddedToEngine) {
+        pEffect->addToEngine(m_pEngineEffectChain, m_effects.size() - 1, m_enabledInputChannels);
+    }
+    // emit(effectChanged(m_effects.size() - 1));
+    slotChainEffectChanged(m_effects.size() - 1);
+}
+
+void EffectChainSlot::replaceEffect(unsigned int effectSlotNumber,
+                                EffectPointer pEffect) {
+    qDebug() << debugString() << "replaceEffect" << effectSlotNumber << pEffect;
+    while (effectSlotNumber >= static_cast<unsigned int>(m_effects.size())) {
+        if (pEffect.isNull()) {
+            return;
+        }
+        m_effects.append(EffectPointer());
+    }
+
+    qDebug() << "Added to engine? = " << m_bAddedToEngine;
+
+    EffectPointer pOldEffect = m_effects[effectSlotNumber];
+    if (!pOldEffect.isNull()) {
+        if (m_bAddedToEngine) {
+            qDebug() << "Removing " << pOldEffect->getManifest()->id() << " loaded at " << effectSlotNumber;
+            pOldEffect->removeFromEngine(m_pEngineEffectChain, effectSlotNumber);
+        }
+    }
+
+    m_effects.replace(effectSlotNumber, pEffect);
+    if (!pEffect.isNull()) {
+        if (m_bAddedToEngine) {
+            qDebug() << "Adding " << pEffect->getManifest()->id() << " at " << effectSlotNumber;
+            qDebug() << "INPUT CHANNELS: ";
+            for (auto channel : m_enabledInputChannels) {
+                qDebug() << channel;
+            }
+            qDebug() << "END INPUT CHANNELS";
+            pEffect->addToEngine(m_pEngineEffectChain, effectSlotNumber, m_enabledInputChannels);
+        }
+    }
+
+    // emit(effectChanged(effectSlotNumber));
+    slotChainEffectChanged(effectSlotNumber);
+}
+
+void EffectChainSlot::removeEffect(unsigned int effectSlotNumber) {
+    replaceEffect(effectSlotNumber, EffectPointer());
+}
+
+void EffectChainSlot::refreshAllEffects() {
+    for (int i = 0; i < m_effects.size(); ++i) {
+        // emit(effectChanged(i));
+        slotChainEffectChanged(i);        
+    }
+}
+
+unsigned int EffectChainSlot::numEffects() const {
+    return m_effects.size();
+}
+
+const QList<EffectPointer>& EffectChainSlot::effects() const {
+    return m_effects;
+}
+
+EngineEffectChain* EffectChainSlot::getEngineEffectChain() {
+    return m_pEngineEffectChain;
+}
+
+void EffectChainSlot::sendParameterUpdate() {
+    qDebug() << "ADDEDD TO ENGINE = " << m_bAddedToEngine;
+    if (!m_bAddedToEngine) {
+        return;
+    }
+    EffectsRequest* pRequest = new EffectsRequest();
+    pRequest->type = EffectsRequest::SET_EFFECT_CHAIN_PARAMETERS;
+    pRequest->pTargetChain = m_pEngineEffectChain;
+    pRequest->SetEffectChainParameters.enabled = m_bEnabled;
+    pRequest->SetEffectChainParameters.mix_mode = m_mixMode;
+    pRequest->SetEffectChainParameters.mix = m_dMix;
```

Comment:
Storing these states in private member variables is redundant with the ControlObjects. We can get rid of this redundant variables and all their setter/getter functions by making this function a slot and connecting it to the `valueChanged` signal of the ControlObjects.

Reply:
I think this should be your next step of cleanup.